### PR TITLE
Include Homebrew bin when searching for Docker binaries

### LINKED ISSUE #3476: [BUG] Docker & Docker compose installed from Homebrew not recognized
**Describe the bug**

I am trying to setup dev environment on Mac M1 and I have docker installed from Homebrew (`brew install docker docker-compose;`). When I start `assemble` task I get an exception about missing Docker binaries like this:

```
Docker (checked [/usr/bin/docker, /usr/local/bin/docker]) is required to run the following tasks: 
```

I think the issue is that `brew` by default installs binaries into different location. See:

```
% which docker docker-compose
/opt/homebrew/bin/docker
/opt/homebrew/bin/docker-compose
```

```sh
# Just FYI my version of Docker-compose and Docker

% docker-compose version
Docker Compose version 2.5.1

% docker version
Client: Docker Engine - Community
 Version:           20.10.16
 API version:       1.41
 Go version:        go1.18.2
 Git commit:        aa7e414fdc
 Built:             Wed May 11 16:22:17 2022
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.14
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.15
  Git commit:       87a90dc
  Built:            Thu Mar 24 01:48:38 2022
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          v1.5.11
  GitCommit:        3df54a852345ae127d1fa3092b95168e4a88e2f8
 runc:
  Version:          1.0.3
  GitCommit:        f46b6ba2c9314cfc8caae24a32ec5fe9ef1059fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

The exception seems to be coming from `DockerSupportService.java`

https://github.com/opensearch-project/OpenSearch/blob/e9b19a095c2cf90274c41aa701432a2748c03468/buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java#L165

It does not currently check for Docker binaries in the location where brew installs it.

**To Reproduce**

A fresh pull from `main`:
```sh
% git pull upstream main
From github.com:opensearch-project/OpenSearch
 * branch                    main       -> FETCH_HEAD
Already up to date.

% git log --oneline -n 1
e9b19a095c2 (HEAD -> main, upstream/main, lukas/main) [Segment Replication] Add SegmentReplicationTargetService to orchestrate replication events. (#3439)

```
Run `assemble` task.

```sh
% ./gradlew assemble    

> Configure project :qa:os
Cannot add task 'destructiveDistroTest.docker' as a task with that name already exists.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 7.4.2
  OS Info               : Mac OS X 12.3.1 (aarch64)
  JDK Version           : 17 (Eclipse Temurin JDK)
  JAVA_HOME             : /Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home
  Random Testing Seed   : 86D4E779628BC837
  In FIPS 140 mode      : false
=======================================

FAILURE: Build failed with an exception.

* What went wrong:
Docker (checked [/usr/bin/docker, /usr/local/bin/docker]) is required to run the following tasks: 
:distribution:docker:buildArm64DockerImage
:distribution:docker:buildDockerImage
you can address this by attending to the reported issue, or removing the offending tasks from being executed.
```

**Expected behavior**

`DockerSupportService.java` should be able to detect Docker tools in location used by `brew`.

**Host/Environment (please complete the following information):**
 - OS: Mac OS,  chip M1: ```% uname -a
Darwin Lukass-MacBook-Pro.local 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64```

### PR DESCRIPTION
**Pull request by lukas-vlcek**
Signed-off-by: Lukáš Vlček <lukas.vlcek@aiven.io>

### Description
When docker and docker-compose are installed using `brew` then the `/bin` location is different (compared to manual binaries installation). This commit extends possible locations that are searched when checking if docker is available on Mac family Os machines.
 
### Issues Resolved
Closes #3476
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java [hunk 1]
```diff
@@ -68,17 +69,30 @@ public abstract class DockerSupportService implements BuildService<DockerSupport
     // Defines the possible locations of the Docker CLI. These will be searched in order.
     private static String[] DOCKER_BINARIES_UNIX = { "/usr/bin/docker", "/usr/local/bin/docker" };
 
+    // On Mac you can install docker and docker-compose via brew adding more options where the binaries can be located.
+    private static String[] DOCKER_BINARIES_MACOS = Stream.concat(
```
Comment (nknize) [inline buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:73]: `/opt` isn't mac specific. Not sure we need the `DOCKER_BINARIES_MACOS`. Maybe make this `DOCKER_BINARIES_HOMEBREW` and add `/opt/homebrew/bin/docker`, and `/home/linuxbrew/.linuxbrew` per the [homebrew documentation](https://docs.brew.sh/Installation)?

**Review by nknize — CHANGES_REQUESTED** [review]
Thanks for doing this. Just left a quick question if we should make this more homebrew specific? Or do we think there will be other macOS specific urls?

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java (hunk 1 above)
Comment (lukas-vlcek) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:73]: @nknize I think this is fair point. Let me update the PR.

Comment (lukas-vlcek) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:73]: @nknize Thinking more about this, depending on how robust we want to make it but why don't we simply use `which docker` and `which docker-compose` on Linux and MacOS based systems instead? My gut tells me there can be always some edge case where docker (or other binary in general) can be installed in custom location...

What is the chance that `which docker` fails (i.e. gives no useful answer) if the `docker` is installed on the system?

If I understand the code correctly then the only change/implication would be that the paths to search in would not be static class members but they will have to be evaluated only after the `ExecOperations execOperations` is injected into the SupportService ctor. That shouldn't be a big deal... WDYT?

Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:73]: I like that idea. How about using the static variables as a fall back if `which docker` does fail? And since OS X is unix based anyway, you should just be able to add the new homebrew paths to the existing `DOCKER_BINARIES_UNIX` variable instead of a new `DOCKER_BINARIES_MACOS`?

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java [hunk 2]
```diff
@@ -66,13 +67,17 @@ public abstract class DockerSupportService implements BuildService<DockerSupport
 
     private static Logger LOGGER = Logging.getLogger(DockerSupportService.class);
     // Defines the possible locations of the Docker CLI. These will be searched in order.
-    private static String[] DOCKER_BINARIES_UNIX = { "/usr/bin/docker", "/usr/local/bin/docker" };
+    private static String[] LINUX_BIN_LOCATIONS = { "/usr/local/bin", "/usr/bin", "/opt/homebrew/bin", "/home/linuxbrew/.linuxbrew/bin" };
```
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:70]: `"/home/linuxbrew/.linuxbrew/bin"` this one is unclear: basically we are pointing to the home folder of the `linuxbrew` user which basically should fail the lookup unless run under `linuxbrew` user or `root`. Or I am missing something?

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java (hunk 1 above)
Comment (lukas-vlcek) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java]: Hi @nknize,

I updated my PR and I think it is ready for review.

I decided to not include the `which` functionality into this PR for now because it turns out to be more complicated. Basically the `which` output can differ based on host systems quite a lot (these are more or less corner cases but we need to deal with them as well, things like "aliases" or "shell functions"). This would lead to more complex parsing of the host output than I like for this PR. I will open a new ticket to enable discussion about it and provide some of the findings I made. In the end I think we will need to add functionality where users will define their own path to container engine anyway (think about people who have both the `docker` and `podman` installed at the same time [yes, I do believe we will have `podman` build compatibility in the future 😄 ]).

For now, let this PR address only the Homebrew locations.

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java (hunk 2 above)
Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:70]: I'm not sure anything will fail here unless the Docker CLI can't be found at all? I think like the `PATH` env variable it will search folders even if they don't exist and only throw an error if the binaries are not present? Might verify? 

Aside from that technical nuance, I think I agree that we should leave the ~linuxbrew user out of the path here. Users can always create a `/usr/local/bin` softlink to the linuxbrew home directory if this is their configuration.

Comment (reta) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:70]: Thanks @nknize , I think "fail" is indeed too strong, I would rephrase that as "unsuccessful lookup" since we attaching to particular user, thank you.

Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:70]: > Perhaps we should add one or two debug messages here as well?

Debug verbosity here would be good! Anything to help users understand what's going on is a great thing. I'll keep an eye on the PR and review when ready.

**Comment by nknize** [comment]
Just checking in on this @lukas-vlcek. I think there was one outstanding item?

> Perhaps we should add one or two debug messages here as well? If I am not mistaken right now user does not know which location is used.

**Comment by lukas-vlcek** [comment]
@nknize Yes, thanks for pinging me. I will be finishing this by the end of the week for sure.

**Comment by lukas-vlcek** [comment]
Interesting, why is github-actions bot trying to use the old commitID?
I did force push into my branch and the new commit should be `2385dd3` and not `cb26cfb`.

**Comment by prudhvigodithi** [comment]
I see [.pull_request.head.sha](https://github.com/opensearch-project/OpenSearch/blob/main/.github/workflows/gradle-check.yml#L13) does not give the force commit sha
Found some related issue.
https://github.community/t/github-actions-how-to-to-get-pr-merge-commit-sha-in-push/209044
@lukas-vlcek can you do a dummy commit and do a normal push again please?
Also just curious did u use `--force-with-lease` or `--force` while pushing ?

**Comment by lukas-vlcek** [comment]
@prudhvigodithi I used just `-f`.

I will push update shortly.

**Comment by bbarani** [comment]
> 

Thanks. The gradle check job has picked up the new commit (https://build.ci.opensearch.org/job/gradle-check/124/consoleFull) now.

**Comment by prudhvigodithi** [comment]
> @prudhvigodithi I used just `-f`
Thanks @lukas-vlcek I have raised this as a [bug](https://github.com/opensearch-project/OpenSearch/pull/3481#issuecomment-1172663112) to investigate more.
Thank you

**Comment by lukas-vlcek** [comment]
Thank you @prudhvigodithi !

File: buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java [hunk 3]
```diff
@@ -98,48 +103,61 @@ public DockerSupportService(ExecOperations execOperations) {
...
-                if (lastResult.isSuccess()) {
-                    version = Version.fromString(lastResult.stdout.trim(), Version.Mode.RELAXED);
-
-                    isVersionHighEnough = version.onOrAfter(MINIMUM_DOCKER_VERSION);
-
-                    if (isVersionHighEnough) {
-                        // Check that we can execute a privileged command
-                        lastResult = runCommand(dockerPath, "images");
+            // Gradle BuildService implementations must be thread-safe
+            synchronized (this) {
```
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:107]: This is interesting, I think we discussed that @lukas-vlcek offline, the `volatile` +  `synchronized` definitely make sense to keep implementation thread safe, I am not sure it is worth doing that here: at worst, the `getDockerAvailability()` may be called a few times, but this is no data race and the outcome would be the same (since the environment is the same).

Comment (reta) [reply buildSrc/src/main/java/org/opensearch/gradle/docker/DockerSupportService.java:107]: @nknize the logic is atomic (the state is a single pointer), the Docker detection logic could run a few times (potentially) which is the only issue I see here, marking `getDockerAvailability()` as `synchronized` would address that (no need for volatile or double check locking) but I have never seen an issue with this service in any builds.
